include/globals.js

```diff
@@ -802,3 +820,69 @@ class ContentResolver {
       }
   }
 }
+
+function HandleTSHStateUpdateMessage(message) {
+  const deltaIdx = message['delta_index'];
+  const deltas = message['delta'];
+
+  if (deltaIdx < maxAppliedDeltaIdx) {
+    throw Error(`Received out of order delta ${deltaIdx} (applied: ${maxAppliedDeltaIdx}`);
+  }
+
+  let event = new CustomEvent("tsh_update")
+  // Ancient jutsu to deep-clone an object. We will be modifying
+  event.oldData = JSON.parse(JSON.stringify(data));
+
+  applyDeltas(data, deltas);
+  event.data = data;
+  maxAppliedDeltaIdx = Math.max(deltaIdx, maxAppliedDeltaIdx);
+  console.log(data);
+
+  // it's fine if this doesn't fire because we threw an exception earlier, because
+  // that means state wasn't applied and we've requested new program state.
+  document.dispatchEvent(event);
+}
+
+/**
+ * Takes a single delta operation, and modifies the data object.
+ *
+ * @param {Object} data
+ * @param {Delta} delta
+ */
+function applyDelta(data, delta) {
```

Comment:
These two functions are ripped straight from the web scoreboard codebase with no changes.